### PR TITLE
[RUM-1830] 128 bit trace IDs

### PR DESCRIPTION
**Pull request by marco-saia-datadog**
### What does this PR do?

Introduces 128 bit trace IDs.

### Review checklist (to be filled by reviewers)

- [x] Feature or bugfix MUST have appropriate tests
- [x] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)
- [ ] If this PR is auto-generated, please make sure also to manually update the code related to the change

### CODE REVIEW

File: packages/core/src/rum/instrumentation/resourceTracking/distributedTracing/distributedTracing.tsx [hunk 1]
```diff
@@ -77,55 +109,177 @@ const generateTracingAttributesWithSampling = (
...
+        return timestamp;
+    }
+
+    /**
+     * Checks if a string representation of an ID in a given radix is within 64 bits.
+     * @param idString - The string representation of the ID.
+     * @param radix - Optional base to use for the conversion (default is 10).
+     * @returns True if the ID is within 64 bits, otherwise false.
+     */
+    static is64Bits(idString: string, radix: number = 10): boolean {
```
Comment (fuzzybinary) [inline packages/core/src/rum/instrumentation/resourceTracking/distributedTracing/distributedTracing.tsx]: I'd rename this function -- `is64Bits` to me means it's exactly 64-bits, which isn't what this is testing really. This is `isLessThan64Bits`.

File: packages/core/src/rum/instrumentation/resourceTracking/distributedTracing/__tests__/distributedTracing.test.ts [hunk 2]
```diff
@@ -9,63 +12,98 @@ describe('TraceIdentifier', () => {
...
+            const idStr128 = id.toString(TracingIdRepresentation.decimal);
 
             // THEN
-            expect(id).toMatch(/[1-9]{1,19}/);
-            // should be less than the max 64 bits integer
-            if (id.length === 19) {
-                expect(id < '9223372036854775807').toBeTruthy();
-            }
+            expect(idStr128).toMatch(/[1-9]{1,19}/);
+            expect(TracingIdentifier.is64Bits(idStr128)).toBe(false);
```
Comment (fuzzybinary) [inline packages/core/src/rum/instrumentation/resourceTracking/distributedTracing/__tests__/distributedTracing.test.ts]: Although unlikely, it is possible for the TraceId to be representable in less than 64-bits if it randomly generates the top 64-bits as zero.

File: packages/core/src/rum/instrumentation/resourceTracking/distributedTracing/__tests__/distributedTracing.test.ts [hunk 3]
```diff
@@ -9,63 +12,98 @@ describe('TraceIdentifier', () => {
...
+    it('M return a valid 64 bits low and high part integer W toString', async () => {
         let iterations = 100;
         while (iterations-- > 0) {
             // GIVEN
-            const trace = new TraceIdentifier();
-            const id = trace.toString(16);
-            const paddedId = trace.toPaddedString(16, 16);
+            const id = new TracingIdentifier();
+            const idStrLow64 = id.toString(TracingIdRepresentation.lowDecimal);
+            const idStrHigh64 = id.highToString(10);
```
Comment (fuzzybinary) [inline packages/core/src/rum/instrumentation/resourceTracking/distributedTracing/__tests__/distributedTracing.test.ts]: Any reason to have `highToString` as a separate method? Couldn't you have `TracingIdRepresentation.highDecimal`?

File: packages/core/src/rum/instrumentation/resourceTracking/distributedTracing/__tests__/distributedTracing.test.ts [hunk 4]
```diff
@@ -9,63 +12,98 @@ describe('TraceIdentifier', () => {
...
+        const currentUnixTime = Math.floor(Date.now() / 1000);
+        const fiveMinutesInSeconds = 5 * 60;
+
+        expect(timestamp).toBeGreaterThan(
+            currentUnixTime - fiveMinutesInSeconds
+        );
+        expect(timestamp).toBeLessThan(currentUnixTime + fiveMinutesInSeconds);
+    });
+
+    it('M return a 128 bits positive hex W toString(16)', async () => {
```
Comment (fuzzybinary) [inline packages/core/src/rum/instrumentation/resourceTracking/distributedTracing/__tests__/distributedTracing.test.ts]:
```suggestion
    it('M return a 128 bits positive hex W toString(TracingIdRepresentation.hex16chars)', async () => {
```

File: packages/core/src/rum/instrumentation/resourceTracking/distributedTracing/__tests__/distributedTracing.test.ts [hunk 5]
```diff
@@ -9,63 +12,98 @@ describe('TraceIdentifier', () => {
...
+            expect(TracingIdentifier.is128Bits(id, 16)).toBe(true);
+        }
+    });
+
+    it('M return a 128 bits positive padded hex W toPaddedString(16, 64)', async () => {
         let iterations = 100;
         while (iterations-- > 0) {
             // GIVEN
-            const id = new TraceIdentifier().toPaddedString(16, 32);
+            const id = new TracingIdentifier().toPaddedString(16, 64);
```
Comment (fuzzybinary) [inline packages/core/src/rum/instrumentation/resourceTracking/distributedTracing/__tests__/distributedTracing.test.ts]: Do we still need `toPaddedString` if we have `toString` taking representations?

File: packages/core/src/rum/instrumentation/resourceTracking/distributedTracing/distributedTracing.tsx [hunk 6]
```diff
@@ -77,55 +109,177 @@ const generateTracingAttributesWithSampling = (
...
-    toString = (radix: number) => {
-        let low = this.low;
-        let high = this.high;
-        let str = '';
+    /**
+     * Generates a 128-bit ID in the format:
+     * <32-bit Unix timestamp> <32 bits of zero> <64 random bits>.
+     * @returns An object containing the full 128-bit ID, high 64 bits, and low 64 bits as hexadecimal strings.
+     */
+    private generate128BitID(): { id: string; high: string; low: string } {
```
Comment (fuzzybinary) [inline packages/core/src/rum/instrumentation/resourceTracking/distributedTracing/distributedTracing.tsx]: Could we possibly generate the ID as a BigInt, which would save re-parsing it when writing out the representations.

File: packages/core/src/rum/instrumentation/resourceTracking/distributedTracing/__tests__/distributedTracing.test.ts [hunk 7]
```diff
@@ -9,63 +12,98 @@ describe('TraceIdentifier', () => {
...
+            const id = new TracingIdentifier();
+            const idStr128 = id.toString(TracingIdRepresentation.decimal);
 
             // THEN
-            expect(id).toMatch(/[1-9]{1,19}/);
-            // should be less than the max 64 bits integer
-            if (id.length === 19) {
-                expect(id < '9223372036854775807').toBeTruthy();
-            }
+            expect(idStr128).toMatch(/[1-9]{1,19}/);
```
Comment (ganeshnj) [inline packages/core/src/rum/instrumentation/resourceTracking/distributedTracing/__tests__/distributedTracing.test.ts]: /blocking

it looks like we are not asserting the id as per spec
 https://datadoghq.atlassian.net/wiki/spaces/RUMP/pages/3545630931/RFC+Support+128+bit+trace+IDs+in+RUM+SDKs#128-Bit-TraceId-Generation

I wonder if we simply parse the string would be simple and maintainable then using a regex.

File: packages/core/src/rum/instrumentation/resourceTracking/distributedTracing/distributedTracing.tsx [hunk 8]
```diff
@@ -77,55 +109,177 @@ const generateTracingAttributesWithSampling = (
...
+    private id: string;
+    private high: string;
+    private low: string;
 
     constructor() {
-        // We need to have a 63 bits number max
-        this.high = Math.floor(Math.random() * MAX_31_BITS_NUMBER);
-        this.low = Math.floor(Math.random() * MAX_32_BITS_NUMBER);
+        const idData = this.generate128BitID();
+        this.id = idData.id;
```
Comment (ganeshnj) [inline packages/core/src/rum/instrumentation/resourceTracking/distributedTracing/distributedTracing.tsx]: /blocking

this is redundant information and can be avoided.

File: packages/core/src/rum/instrumentation/resourceTracking/distributedTracing/distributedTracing.tsx [hunk 9]
```diff
@@ -77,55 +109,177 @@ const generateTracingAttributesWithSampling = (
...
-     * @returns padded string
+     * Returns a string representation of the TracingId.
+     * @param representation - The type of representation to use.
+     * @returns The ID as a string in the specified representation type.
      */
-    toPaddedString = (radix: number, length: number) => {
-        const traceId = this.toString(radix);
-        if (traceId.length > length) {
-            return traceId;
+    public toString(representation: TracingIdRepresentation): string {
```
Comment (ganeshnj) [inline packages/core/src/rum/instrumentation/resourceTracking/distributedTracing/distributedTracing.tsx]: /question

expect this, all other APIs that are below private non? If yes, can we mark them private?

File: packages/core/src/rum/instrumentation/resourceTracking/distributedTracing/distributedTracing.tsx [hunk 10]
```diff
@@ -77,55 +109,177 @@ const generateTracingAttributesWithSampling = (
...
  */
-export type TraceId = TraceIdentifier & {
+export type TraceId = TracingIdentifier & {
     _brand: 'traceId';
 };
 
-export type SpanId = TraceIdentifier & {
+export type SpanId = TracingIdentifier & {
     _brand: 'spanId';
 };
```
Comment (ganeshnj) [inline packages/core/src/rum/instrumentation/resourceTracking/distributedTracing/distributedTracing.tsx]: /concerning

this is slippery slope that we are sharing the same type here which increases chance of changing the behavior of the `SpanId` accidentally. This was one of the reasons we went with separate types in iOS for SpanId and TraceId.

Somewhere down the line we are using
```
   {
                        header: PARENT_ID_HEADER_KEY,
                        value: tracingAttributes.spanId.toString(
                            TracingIdRepresentation.lowDecimal
                        )
                    },
```

which is not great as `PARENT_ID_HEADER_KEY` is a `SpanId` always.

File: packages/core/src/rum/instrumentation/resourceTracking/distributedTracing/distributedTracing.tsx [hunk 11]
```diff
@@ -77,55 +109,177 @@ const generateTracingAttributesWithSampling = (
...
+    public lowToPaddedString(radix: number, length: number): string {
+        return BigInt(`0x${this.low}`).toString(radix).padStart(length, '0');
+    }
+
+    /**
+     * Extracts the Unix timestamp from the 128-bit hex string representation.
+     * @param idHex - The 128-bit ID as a hexadecimal string.
+     * @returns The Unix timestamp as a number.
+     */
+    public static extractTimestamp(idHex: string): number {
```
Comment (ganeshnj) [inline packages/core/src/rum/instrumentation/resourceTracking/distributedTracing/distributedTracing.tsx]: /improvement

this and below looks like test only code, can we move them to __tests__?

**Review by ganeshnj — COMMENTED** [review]
Some suggestions, overall on right track. 

One thing that I couldn't see how we are dealing with RUM, Trace or Log events with this update. I assume some encode/decode to be changed.

**Comment by marco-saia-datadog** [comment]
I have changed the implementation, here is how it works now:

## Separating Span ID and Trace ID

The `TracingIdentifier` constructor is now private.

It can only be created through `TracingIdentifier.createSpanId()` or `TracingIdentifier.createTraceId()`, and it returns different types respectively.

Span IDs will always be generated as 64 bits BigInt(s) and Trace IDs as 128 bits BigInt(s)

## String representation takes into account the type of Tracing Identifier

TracingIdFormat (previously `TracingIdRepresentation`) is now more generic, and the `.toString` implementation has been modified accordingly.

When requesting a padded HEX value, the bit length of the current type will be taken into account.

Requesting the low bits of a 128 bit Trace ID will get you a 64 bit ID, while requesting the low bits of a 64 bit Span ID will get you a 32 bit ID.

Here is a [runnable example](https://www.typescriptlang.org/play/?target=99#code/PQKhCgAIUhBA3AhgSwDaIEaoKaQGYD2ATgLaIAuAzvsZEdgA72XYB25yrA5pOQBa4A3gAFUnANaQAKkUQBjTlwCSAEzYc8ybEQC+kRNUSRK5IooB0UEMHBsAriWmyF3VQDFiZcpEFRI-0Ah-YJgAEWwFMlRjU0U6RmZ1CmQCVkgCPF4BfDtUaNN5OOQVS2D-az9-NUjEVAAacErIQKboaWzUAgB3SAxkKnTM-lwClx5i-UNIauQo0rKKss6u8Jr6xrKWsrapbL5kLj5e-uoMrJHnIpVJ-WmI2dr5kJsy-cPVh-WmrYXIAAlsAAPRAzKIxMzceJMbAsdjJVKDc45PK8S6Q4pPcovYICQENb5gVowAHA0G1cFxejQ2HkeFpM7DSDLY4DBnZUZXTHQbH+ZYk-GbQnbYlAkH3MEmCE8KmJOEcBFs3BvI59VlDdlo8YlIk8yDK-kbYI-EKQAAKILU11xYrWFMhMphSXl9PVuDwuXymsgGJ1TQYFuwKgNBKCv3NKktetFZOikspCUdcpSLqRzNVp1dqMK6O1wt1-ojgYAMt1g4LQybw5HrTG7dKEzS6YjGcqWRmkRyc1zFsEC5a-gc+CTwDpGlsYLAfKIJE5s8o1OxkJptHo7CxroQiLO5DDIAAKACMACYABwsgCUpQq5AAngxcDJ5NhVJAALzbxSqdRLrRbgBkPhNAA+hgsisCoABckAAOSjM+KjQQA3COyHgOOcBTmIrCSI+Yxfouy66JAa6BjQW4AMr+qw1B7gAbAALBeV42Le96QJRiCsC+764Z+C4aL+kAAb4ZQgWBkEwZQVGqEhKFjkKMC7Bcc7evxP7aLwd7YMx4CsQ+mr4QJ2hSFpb4wXB0GQAAPpJVGyWhCmQAAarUdi4CR1zkAQkBkJQkiMsyDFts2GpPqp36ETpcipCYkBFgA8gA6kBDEAEJKFIQEALKwORADSZl7qlBxKOwe7QQADICeA1bVNXQeekAADxNZAADMR6sI1ADUkDFcoZWVdVdW1Q1qHoS5qBucR668N5vn+R03TtUewWKlmO7hQRv5RTF3gJclHXpZlOX5YV-WleQ5VVXVDXNa1B60V1kC9Rdg03aN57jY5k4iFhOEGWphF6MgUx2KwyAAI7TQAqjDSihFkFCQHInG9LgDGqukW7Hieqp1Po4GQEwBDwMUMJWCjqTwGwWjsD52D8AQKjUF5UKyt4-Q3EYfwAKIABrY5AoS8wAwkoOVFnWBOcdcXO1MsrPZAtiiU2cJxMsttCtumV6U6L9AULgRisNgPTk9tGkYDeKMK3Ef0zrxrhAztciG+Q2C4fBe7nnoxCUw72Efs7EWu+72AcVxKg+zoUXoJQ1BO-OocaSJRpCr8ABK2AggAtKkqA21wbDaEb1wI9zmEzn0XCcOQse+mUhsqAXNvFFBNd16h5ZEpA2d563mlsWzage6QnC4F0AjDFujIW0ZW6g7cgcA0+qh+1uRgr+x0kqA3eZNM3g96VBSeGepRAmfe3fpxW5SQAbOce1Mps9ODUPTV7kAI92kDCPQ5A7BEGokiYuptZAe2uNvL269uz5jsFgZAcgYjJGQW7J+ntnDe3PKfLBL407bHiIA4BkBX7B2TpbIg5ULKNQMNueCyEyijhDL3R+RsX5m2IhDaGuBI7f1CL-f+jMgEgMZGA0ukCq5B0jrAxuvYEFiGQSYVBKNw4yOjjgnenF8GtGCAAkRpDOFnxdtocqUlOJ3ToeoxhwRmE92FGaMwSAPZU2oqYOwcgvKz28pwfoyBajIAAF4jGyNAwGKciIYAMKRBEYjkA01YL3MJc5z6ESvtgUcB8yhMHiUbVxkoPFeL3CfchqTfzpMagQ7YlAuj9DkEcYpWlKm6MIajFg5ksHQQgi0whwR+Cg3MBMd8-TKDmHERA7AcMEblVxqqMaPTem9ENuIGxiz-BtNwNBcxrAukLN6SMwZ1xhn7FGeMo2UzQjlUxv0eZazemgRzisnpdj9knPMHpMyelVmQBeQEDOJoADiJcJlg24dNCutT+CgPiWwMiXhBH+lkI4TcXhIC5yRGoSgyB6Abk8MjOi9Esa0Fmf0S8vchHENEdkM5kiLmV23p3dg+8Fj5icXkmlkz4aXJRRQKC0FrnkEsjZaCJLBWaMZd4KpmxgCQCBd4RkcggH0HpjDCGgJeCzBhLSEgDBvRpBYNFcClAemGtiu-QE5EIipBZmZLKFA+DmDwJ0Yge5QhG3MKwboPtmiQAPBVf1X1DSEOADK3m1EgEhNwKq5A6qOAkC1YgHV3pqAdTbCavaXCY2WsNSzVN75zXZutdQACH0Rp4BvsGmVqb0yImCUQAg6a3GQDrQQSgebIAVQrdsENsrgV5KCuJAgjga1rjiHa-g5hB0kB9o22KU6jr9APLa+1jrnVUPHQ6qd3qYClpGoGxZprvDzqPMVcgq13wbtXQQF1l6t2NR3cNMt+7emHroLLIdDESRvj2f4Y9p6DzvIIORWI3BDy0UvAWYDiAiBXRPATSqd1uo-rfeBIdC6z2AeA1KMDEGQRQZg3uODMEKrzJ6T2pQmQeWc2oAKyAXRcD6JIVzWY8aVD+I9oXHpS59xUbfK+d8-LCU3OaXc39wiSFvSugAAyqgAEkEFOz9QIdBSefYQ353aZWi2ptobwiKBhsxrOKckcZuDGoPRmgtVqjVfvzWqwtRqOqYZA1wHD5hIO0gI0RhDantivpbZQWzzbtCtqc15LDig3Meeg7B+DJHA1ka00OvoptiYxdZt5FFSJca5yxgjWd3hcS41PRXOzWbrMsy-b1ALVWUMtxIEpwEXam7ibSJJvcMnATyaK6eEroQVO+Y08ae+2dKVTFM-WakTpkwhX0ik3MvxhCIsTezRMtJnRovaCMUyZwHSNg22zEigjGMgKUvwyuE29VIikvcZc1w9vTdiVpOBfoFFILmhF0DD2kypFwfNjwpAKCaMu1K4Ir7lheztX5b9omsgDI+fxgTFlIAAH44pJRSvRY62VcoFSggdICR0Mo4-yt8sor7lSQ4MJIY58PTKI46U+SyaOGJpCgh1VgZOwcZr7HEWnoyEf8cZzuZnK1IBQUes14INS6kNO++t5MInRMbNKSoAH8KYzdNhy1ylcPRkYnCy5w8FUEuw5V8Y9XFBzDLA+FELX2uxO6+KW8iYAEIdYKh+IS8hvsN+tN8r6JqvLfkHMMqW3tR7cO5O-uA5EwAB8ce9SDip35b3QGjd+6l60wPFu8Uh9xJH7X0fY8lB95Fx6-u7nm8BsH63pYgSF9h9H53AzXeay6Cnr3znffgaz35nPNe8+h8HCSRvonm8l8gAnpPhxO9p8+65ivffycD-+0PvsgZR-IZ1wYg5ZfQO58B-noEuGVD4autv3pvPuANAd4QhDl--CV7WdXtfR-3MBhUCWLoW+79EN328vvq5ofvCnyCfh-mfp5hfn-tkhaHEDKkeLfjAf4A-n-s-geqvnhGruvp-gOIcL-nfsXoAenthiAVbvqOAdFgRo-tsNfjwAgUgcgagXfr5kwk0KOMwmhDKrnDwbwXwfwTwVwb0nzPzEIQIeIXweAK+nBNxKrhEuYOgkbDAhouNDKjDAAHKizxRZRZS8zqFSCQBuDxSZzSC8zkRSBKDqEArdI9oyElBDIdqc6NDSFYJBZ2Hd6RZkHH6AiBouFPgkh4FHDDJ4Kl4kGeGD7v4UE+HIR+E7gkjf6fIhEeEH4RGgH17RHOExQEA4B16ua+FZE5GdCuZSZfwVxuAwxFhSwiFQTyZwQkgDYxEFHaRFEdalGIx-BKAAp-D-ACw1GCB1FAiBENFSFNG5GtFYLnYHQ9H8x9EDGAjf7DEjFNpwThhuFJFAF7heEQH9gn6NHLFYKrGDGDiJFryhEL6bGpFW4b5Bgj67FLGxQrEgjxHLTBGnHJHAGXEh7XHf4kgJamrZHNEEB5H3EAljElETEVymiwChAiyIzlGVHTGzEHFPHKaqZ7GUCgktHglhSQnQmwn-CdHdHVGQC1HIlBhHGHDDH-GFFAnjE4mIxQkwm8yIxTHEmklPiHHzHdCLGvrbKyHGLyGKEezqI+yqGQAaFaE6F6EGFGEmFSBmEWFWE2Eyp8n2FHKOGoS8lURBaqnvEXFv7wq4j5FNrbIBHHHvi6kbFbFRHGmxSmlAgJEWm7x6lbFgEZEgk0l5HomYm0lSZ8JlEVFVG9EkmCD2mAhUmjFYn+ntGEmIkhlhlDFokemAnFHRno6JRxnxnakOnclJmZEmlUSclmSWlhEpEGlXG4F3FamcSrHmnGDOlWmfHbGb63EZHVmsDhiOn1naJnFG5bHfHpF-GRm0nnjJlglpmMn4nwlBkzFZk1konhlJnUkpkdYTl4nMkEldFxnybbK1mUlLnDmplUTnaTkbmsnBk7mFkggLF5k9oSH3lCHbDh6oBiH3niH3HypYKrAnE7iqAulNkxi2mflPirCBE-nwT-nln56DjPlAUbTYCrBdnuGNlQV14rDGaoBDnUQ+leljlYltGGGBnCxixIkgURARnYWel0mbQVwdFbkiyiykU7igWDgUUYlUXYk0UslJTEWMUhlwSIW5lYV2lUTflOk9mQVYG16AXoneDbIsWHDFkNmlkfGoVh4YVwXyURBdklnnGundCwWamHmil4W+lpnTm8V9FaVyBsU4WrnHm0WxkMVWWiURCJm2l2V+kOXcUZnOVzmsCCVdANFAA).

Here is an example output for reference:

| FORMAT                    | VALUE                            |
|---------------------------|----------------------------------|
| Trace ID FULL HEX         | 667ed836000000000d8dbaedc8adcf59 |
| Trace ID PADDED FULL HEX: | 667ed836000000000d8dbaedc8adcf59 |
| Trace ID HIGH HEX         | 667ed83600000000                 |
| Trace ID PADDED HIGH HEX: | 667ed83600000000                 |
| Trace ID LOW HEX          | **d8dbaedc8adcf59**              |
| Trace ID PADDED LOW HEX:  | **0d8dbaedc8adcf59**             |
| Span ID FULL HEX          | 378ad79b0519da05                 |
| Span ID PADDED FULL HEX   | 378ad79b0519da05                 |
| Span ID HIGH HEX          | 378ad79b                         |
| Span ID PADDED HIGH HEX   | 378ad79b                         |
| Span ID LOW HEX           | **519da05**                      |
| Span ID PADDED LOW HEX    | **0519da05**                     |

I will add a few more tests to match the RFC specs more accurately, and verify @ganeshnj's comment https://github.com/DataDog/dd-sdk-reactnative/pull/680#pullrequestreview-2141204325.

**Comment by marco-saia-datadog** [comment]
Example of generated headers:

| Header                        | Value                                             |
|-------------------------------|---------------------------------------------------|
| x-datadog-origin              | rum                                               |
| x-datadog-sampling-priority   | 1                                                 |
| x-datadog-trace-id            | 17731188736533748218                              |
| x-datadog-parent-id           | 17034905968583185706                              |
| x-datadog-tags                | _dd.p.tid=6682b2ee00000000                        |
| traceparent                   | 00-6682b2ee00000000f611d636468059fa-ec6824b93f95152a-01 |
| tracestate                    | dd=s:1;o:rum;p:ec6824b93f95152a                   |
| b3                            | 6682b2ee00000000f611d636468059fa-ec6824b93f95152a-1 |
| X-B3-TraceId                  | 6682b2ee00000000f611d636468059fa                  |
| X-B3-SpanId                   | ec6824b93f95152a                                  |
| X-B3-Sampled                  | 1                                                 |

File: packages/core/src/rum/instrumentation/resourceTracking/distributedTracing/TracingIdentifier.tsx [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+        return new TracingIdentifier('span') as SpanId;
+    }
+
+    /**
+     * Private constructor to initialize the {@link TracingIdentifier} based on the given
+     * {@link TracingIdentifierType}.
+     */
+    private constructor(type: TracingIdentifierType) {
+        switch (type) {
+            case 'trace':
```
Comment (ganeshnj) [inline packages/core/src/rum/instrumentation/resourceTracking/distributedTracing/TracingIdentifier.tsx]: /improvement

you can extract the enum values to constants.

File: packages/core/src/rum/instrumentation/resourceTracking/distributedTracing/__tests__/distributedTracing.test.ts [hunk 13]
```diff
@@ -9,63 +17,327 @@ describe('TraceIdentifier', () => {
...
+                const id = TracingIdentifier.createTraceId();
+                const idStr128 = id.toString(TracingIdFormat.hex);
 
-            // THEN
-            expect(id).toMatch(/[1-9]{1,19}/);
-            // should be less than the max 64 bits integer
-            if (id.length === 19) {
-                expect(id < '9223372036854775807').toBeTruthy();
+                // THEN
+                expect(idStr128).toMatch(/^[0-9a-f]{1,}$/);
```
Comment (ganeshnj) [inline packages/core/src/rum/instrumentation/resourceTracking/distributedTracing/__tests__/distributedTracing.test.ts]: /improvement

Although, I would have preferred a simple check with string parsing which is easy, fast and maintainable however I would let you pick.

if it has to regex, let's make it more strict with something like
```
^[0-9a-f]{8}[0]{8}[0-9a-f]{16}$
```

**Review by ganeshnj — APPROVED** [review]
Nice updates, just a few improvement comments but looks good to me.

**Pull request merged**